src/GobanCanvas.ts

```diff
@@ -1140,54 +1196,61 @@ export class GobanCanvas extends GobanCore {
             }
         }
     }
-    public drawSquare(i: number, j: number): void {
-        if (i < 0 || j < 0) {
-            return;
-        }
-        if (this.__draw_state[j][i] !== this.drawingHash(i, j)) {
-            this.__drawSquare(i, j);
-        }
-    }
-    private __drawSquare(i: number, j: number): void {
```

Comment:
I don't think we can get rid of the drawing hash system, doing some testing on OGS it results in a very notable increase in display latency when you're paging through moves, specifically advancing 10 moves at a time the latency on my machine went from 100ms to about 400. There are likely other use cases where there's degragation too.

Was there a particular challenge you were running into with the hash system? I don't particularly like it, but the benefits of it are pretty apparent so unless we can replace it with something better, we should keep it.

Reply:
To get an extended draw area for the square, I draw the surrounding squares to mix them without glitches. This allows nice shadows and effects like jostled stones, etc. and puts Goban on par with Sabaki.

DrawingHash() would skip these needed draws, so it was a challenge.

I do "kind of" hash inside redraw(), using json.stringify(). I compare the board state to its previous state and only draw a square if different. Full redraw()s were the bottleneck so I moved the "hash" idea there. I also figured stringify() would be faster because it's native.

Anyway, the current lag is my fault; I was planning to gather all needed stone draws and calculate a clip at once for "big redraws", instead of drawing squares over and over, but didn't get to it.

Now I have, in [6b60db3](https://github.com/online-go/goban/pull/63/commits/6b60db36fcb2410907ff8159ec0c002e00f9b5d9). It's a pretty big speedup for me, try it out. It's pretty close to ideal (only draws stones that need it), so if it's still slow I'll have to whip out some other strategy like limiting the extended clip box to only shadows or something.

Reply:
Awesome, that brought it down to be about as fast as the hash system, it's still maybe 25% slower but it's not that noticeable, and you're doing more so I'm fine with that. Though there's some quirks to work out:

![image](https://user-images.githubusercontent.com/168460/158370186-ca5911aa-3973-4049-a136-a6440962a666.png)

![image](https://user-images.githubusercontent.com/168460/158371559-670016ea-f9ce-458c-9d18-453f0eca399b.png)

You can test this if you have your online-go.com repo pointed to the beta (the default) and go to

http://localhost:8080/game/9594

and going to move 36

Reply:
labels weren't getting redrawn because I only drew when a stone was there. Now it draws when label or stone exists.
Fixed with [4b2a81c](https://github.com/online-go/goban/pull/63/commits/4b2a81c4ddab87afde9e909874b896943e8bd5df)
Also, faded lines are drawn in the right order and look less intrusive

Reply:
I improved it a little by paying attention to theme's bounding box. For simple themes (plain, rendered) the draws are now marginally faster because they don't overlap like a JSON theme.
I also put the last move marker in its own little element so as not to keep drawing stones beneath it. 
Finally, JSON themes won't draw shadows when radius < 10. Took me a day + some tea to realize what you meant, sry.
This is in [61b07ca](https://github.com/online-go/goban/pull/63/commits/61b07ca8d207c42da7560e9a08d1554be6985eb6)